src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp

```diff
@@ -0,0 +1,212 @@
+#define ARBORX_DETAILS_DISTRIBUTED_TREE_NEAREST_HELPERS_HPP
+
+#include <ArborX_AccessTraits.hpp>
+#include <ArborX_Box.hpp>
+#include <ArborX_DetailsHappyTreeFriends.hpp>
+#include <ArborX_LinearBVH.hpp>
+#include <ArborX_Point.hpp>
+#include <ArborX_Ray.hpp>
+#include <ArborX_Sphere.hpp>
+
+namespace ArborX
+{
+namespace Details
+{
+
+template <class Predicates, class Distances>
+struct WithinDistanceFromPredicates
+{
+  Predicates predicates;
+  Distances distances;
+};
+} // namespace Details
+
+template <class Predicates, class Distances>
+struct AccessTraits<
+    Details::WithinDistanceFromPredicates<Predicates, Distances>, PredicatesTag>
+{
+  using Predicate = typename Predicates::value_type;
+  using Geometry =
+      std::decay_t<decltype(getGeometry(std::declval<Predicate const &>()))>;
+  using Self = Details::WithinDistanceFromPredicates<Predicates, Distances>;
+
+  using memory_space = typename Predicates::memory_space;
+  using size_type = decltype(std::declval<Predicates const &>().size());
+
+  static KOKKOS_FUNCTION size_type size(Self const &x)
+  {
+    return x.predicates.size();
+  }
+  template <class Dummy = Geometry,
+            std::enable_if_t<std::is_same_v<Dummy, Geometry> &&
+                             std::is_same_v<Dummy, Point>> * = nullptr>
+  static KOKKOS_FUNCTION auto get(Self const &x, size_type i)
+  {
+    auto const point = getGeometry(x.predicates(i));
+    auto const distance = x.distances(i);
+    return intersects(Sphere{point, distance});
+  }
+  template <class Dummy = Geometry,
+            std::enable_if_t<std::is_same_v<Dummy, Geometry> &&
+                             std::is_same_v<Dummy, Box>> * = nullptr>
+  static KOKKOS_FUNCTION auto get(Self const &x, size_type i)
+  {
+    auto box = getGeometry(x.predicates(i));
+    auto &min_corner = box.minCorner();
+    auto &max_corner = box.maxCorner();
+    auto const distance = x.distances(i);
+    for (int d = 0; d < 3; ++d)
+    {
+      min_corner[d] -= distance;
+      max_corner[d] += distance;
+    }
+    return intersects(box);
+  }
+  template <class Dummy = Geometry,
+            std::enable_if_t<std::is_same_v<Dummy, Geometry> &&
+                             std::is_same_v<Dummy, Sphere>> * = nullptr>
+  static KOKKOS_FUNCTION auto get(Self const &x, size_type i)
+  {
+    auto const sphere = getGeometry(x.predicates(i));
+    auto const distance = x.distances(i);
+    return intersects(Sphere{sphere.centroid(), distance + sphere.radius()});
+  }
+  template <
+      class Dummy = Geometry,
+      std::enable_if_t<std::is_same_v<Dummy, Geometry> &&
+                       std::is_same_v<Dummy, Experimental::Ray>> * = nullptr>
+  static KOKKOS_FUNCTION auto get(Self const &x, size_type i)
+  {
+    auto const ray = getGeometry(x.predicates(i));
+    return intersects(ray);
+  }
+};
+
+namespace Details
+{
+
+template <typename Tree, typename Callback, typename OutValue, bool UseValues>
```

Comment:
The `OutValue` argument is really annoying. Currently, no avoiding it. The issue is that we can't determine from the callback's signature what the output value type is going to be. The way we are doing things right now is that determine that from the storage arrays passed in by a user.